test/gurka/test_chinese_postman.cc

```diff
@@ -0,0 +1,574 @@
+  gurka::assert::raw::expect_path_optional(route, expected_names);
+}
+} // namespace
+
+// parameterized test class to test all costings
+class ChinesePostmanTest : public ::testing::TestWithParam<std::string> {
+protected:
+  static gurka::map chinese_postman_map;
+  static gurka::map complex_chinese_postman_map;
+
+  //    A------B---<--C--->--G
+  //    |      |      |      |
+  //    |      |      ^      v
+  //    |      |      |      |
+  //    |      |      |      |
+  //    D------E------F--<---H
+
+  static void SetUpTestSuite() {
+    // Setup chinese_postman_map
+    const std::string ascii_map = R"(
+    p---------q-------r------s-----t
+    |    A----|--B----|--C---|--G  |
+    | i--|--j-|--|--k |  |   |  |  |
+    | |  |  | |  |  | |  |   |  |  |
+    | l--|--m-|--|--n |  |   |  |  |
+    |    |    |  |    |  |   |  |  |
+    |    D----|--E----|--F---|--H  |
+    u---------v-------w------x-----y
+    )";
+    // AB_2 means AB road is a two way road.
+    const gurka::ways ways = {// two ways
+                              {"AB", {{"highway", "residential"}, {"name", "AB_2"}}},
+                              {"AD", {{"highway", "residential"}, {"name", "AD_2"}}},
+                              {"BE", {{"highway", "residential"}, {"name", "BE_2"}}},
+                              {"DE", {{"highway", "residential"}, {"name", "DE_2"}}},
+                              {"EF", {{"highway", "residential"}, {"name", "EF_2"}}},
+                              // one way
+                              {"CB", {{"highway", "residential"}, {"name", "CB"}, {"oneway", "yes"}}},
+                              {"FC", {{"highway", "residential"}, {"name", "FC"}, {"oneway", "yes"}}},
+                              {"CG", {{"highway", "residential"}, {"name", "CG"}, {"oneway", "yes"}}},
+                              {"GH", {{"highway", "residential"}, {"name", "GH"}, {"oneway", "yes"}}},
+                              {"HF",
+                               {{"highway", "residential"}, {"name", "HF"}, {"oneway", "yes"}}}};
+    const auto layout = gurka::detail::map_to_coordinates(ascii_map, 10);
+    // Add low length limit for excluded_polygons so it throws an error
+    chinese_postman_map = gurka::buildtiles(layout, ways, {}, {}, "test/data/gurka_chinese_postman",
+                                            {{"service_limits.max_exclude_polygons_length", "1000"},
+                                             {"service_limits.max_chinese_polygon_length", "1000"}});
+
+    // TODO: this is kinda of a bad graph for testing as there's more than one solution for e.g. F ->
+    // E, when covering the whole graph, it could be either one of these two and both are exactly
+    // equal in cost: { "FD", "DE", "EA", "AB", "BC", "CD", "DE", "EA", "AF", "FE", "EA", "AC", "CD",
+    // "DE" } { "FD", "DE", "EA", "AB", "BC", "CD", "DE", "EA", "AC", "CD", "DE", "EA", "AF", "FE" }
```

Comment:
this also took a while to understand. when I refactored a bit, some tests were failing using this map. when I went through edge-by-edge I realized they're the exact same.. I'm sure it's deterministic, but it definitely depends somehow on the order of edge insertion into the chinese graph or so

Reply:
Uff, those maps are _very_ flaky.. Even in the test code it's expected to have different results for OSX and Linux. Now OSX is failing with what _seems_ like legitimate problems on first glance, but Linux is fine. If possible it'd be very good to have _only_ **one** possible solution, though I realize that might not be too easy and maybe even impossible, there often are multiple equivalent solutions/paths, probably all the more when the map gets more complex. I think there should be more lower level testing of functions rather than so many end-to-end tests which have the aim to test code of lower-level functions.